src/pynwb/icephys.py

```diff
@@ -319,15 +335,492 @@ def __get_row_ids(self, sweep_number):
+                        categories_arg.insert(0, dynamic_table_arg[-1].name)
+                if required_dynamic_table_given[1] < 0:
+                    dynamic_table_arg.append(IntracellularStimuliTable())
+                    if not dynamic_table_arg[-1].name in categories_arg:
+                        categories_arg.insert(0, dynamic_table_arg[-1].name)
+                if required_dynamic_table_given[0] < 0:
+                    dynamic_table_arg.append(IntracellularElectrodesTable())
+                    if not dynamic_table_arg[-1].name in categories_arg:
+                        categories_arg.insert(0, dynamic_table_arg[-1].name)
+            kwargs['category_tables'] = dynamic_table_arg
+            kwargs['categories'] = categories_arg
+
+        call_docval_func(super().__init__, kwargs)
+
+    @docval({'name': 'electrode', 'type': IntracellularElectrode, 'doc': 'The intracellular electrode used'},
+            {'name': 'stimulus_start_index', 'type': 'int', 'doc': 'Start index of the stimulus', 'default': -1},
+            {'name': 'stimulus_index_count', 'type': 'int', 'doc': 'Stop index of the stimulus', 'default': -1},
+            {'name': 'stimulus', 'type': TimeSeries,
+             'doc': 'The TimeSeries (usually a PatchClampSeries) with the stimulus',
+             'default': None},
+            {'name': 'response_start_index', 'type': 'int', 'doc': 'Start index of the response', 'default': -1},
+            {'name': 'response_index_count', 'type': 'int', 'doc': 'Stop index of the response', 'default': -1},
+            {'name': 'response', 'type': TimeSeries,
+             'doc': 'The TimeSeries (usually a PatchClampSeries) with the response',
+             'default': None},
+            {'name': 'electrode_metadata', 'type': dict,
+             'doc': 'Additional electrode metadata to be stored in the electrodes table', 'default': None},
+            {'name': 'stimulus_metadata', 'type': dict,
+             'doc': 'Additional stimulus metadata to be stored in the stimuli table', 'default': None},
+            {'name': 'response_metadata', 'type': dict,
+             'doc': 'Additional resposnse metadata to be stored in the responses table', 'default': None},
+            returns='Integer index of the row that was added to this table',
+            rtype=int,
+            allow_extra=True)
+    def add_recording(self, **kwargs):
+        """
+        Add a single recording to the IntracellularRecordingsTable table.
+
+        Typically, both stimulus and response are expected. However, in some cases only a stimulus
+        or a resposne may be recodred as part of a recording. In this case, None, may be given
+        for either stimulus or response, but not both. Internally, this results in both stimulus
+        and response pointing to the same timeseries, while the start_index and index_count for
+        the invalid series will both be set to -1.
```

Comment:
1. This behavior (in which a PatchClampSeries appears to exist but does not, in fact) was quite confusing and at times facilitated bugs in my code. If possible, not having this artificial reference would be easier for a user, I believe. Or `IntracellularRecordingsTable` should be generated with `None`s when those indices are -1.

2. Another confusion comes from the fact that, if either the `start_index` or the `index_count` is -1, the other one must also be -1 (I don't see how one can be positive while the other one -1). But this relation is never (?) enforced or clearly specified, so a user may wonder if they have to check for both values or one would be enough.

Reply:
Thanks for the helpful feedback. 

>     1. If possible, not having this artificial reference would be easier for a user,

I agree. Unfortunately HDF5 does not a Null value and NAN is not defined for integer types. In the file itself, we therefore unfortunately can't get around storing the data with -1 values. That being said, replacing the -1 values on read with None (i.e, when slicing into the table or when converting it Pandas) should be doable. I'll take a look at this next week. 

>     2. Another confusion comes from the fact that, if either the `start_index` or the `index_count` is -1

I think this is similar to the issue you raised in 1. I think what we need to do is: 1) update the slicing and conversion to pandas for the ``TimeSeriesReferenceVectorData`` to convert entries that have -1 as start_index to None values, 2) update the docs of ``TimeSeriesReferenceVectorData`` to describe this behavior. The -1 value is here again used as a way to indicate None. The reason we need this here, is because our stimulus and response columns must align.

For the  ``add_recording`` function, I think we may also want to change the default value of start_index and index_count parameters to None.  The ``add_recording`` function calls ``__compute_index`` to determine the correct start_index and index_count for the given TimeSeries to use. I.e., if, e.g., not stimulus is given then both ``start_index`` and ``index_count`` will be automatically set to -1. Similarly, if a TimeSeries is provided for a stimulus and ``start_index`` and ``index_count`` were set to -1 then they will be converted by ``__compute_index`` to the appropriate values for the given TimeSeries, i.e, start_index will be set to 0 and the index_count to the length of the TimeSeries. 

>     But this relation is never (?) enforced

It is enforced when calling ``add_recording`` but it should also be enforced in ``TimeSeriesReferenceVectorData`` and the docs should be updated to mention this as well.

Reply:
@lvsltz I have addressed this issue in https://github.com/NeurodataWithoutBorders/pynwb/pull/1383. When creating a file you can now use "None" for missing stimulus/response and when reading the data missing values are masked using numpy Masked arrays. As such, a user should no longer see the -1 values when accessing the data via PyNWB. 

BTW, I am mostly finished with https://github.com/NeurodataWithoutBorders/pynwb/pull/1383 and https://github.com/hdmf-dev/hdmf/pull/645 (mostly adding tests and documentation). Once those are complete things should be ready for release. I expect that we should have all of this released in HDMF and PyNWB by the end of this month.